Lib/difflib.py

```diff
@@ -1047,22 +1032,28 @@ def _qformat(self, aline, bline, atags, btags):
         '+ \tabcdefGhijkl\n'
         '? \t ^ ^  ^\n'
         """
+        def _keep_original_ws(s, tag_s):
+            def _decide(c, tag_c):
+                if not tag_c.isspace():
+                    return tag_c
+                # preserve tab indentation in original string
+                elif c.isspace():
+                    return c
+                else:
+                    return ' '
+
+            return ''.join(_decide(c, tag_c) for c, tag_c in zip(s, tag_s))
 
```

Comment:
The only kind of whitespace in a tag string is a blank, so this can be replaced with the simpler (and, to my eyes, clearer) clause
```python
    c if tag_c == " " and c.isspace() else tag_c
```
Then there's no need for a nested `_decide()` function (or the expense of calling it many times), and the body of `_keep_original_ws()` can be one statement:
```python
    # Replace blanks in tag_s with the corresponding whitespace
    # characters (if any) in `s`.  The aim is to improve visual
    # output when `s` contains tabs.
    return ''.join(c if tag_c == " " and c.isspace() else tag_c
                   for c, tag_c in zip(s, tag_s))
```
It would also be good then to move this private function to module level, (as the old `_count_leading()` was), so we don't incur the expense of creating a new function each time `_qformat()` is called.

Reply:
heh, I had that originally but thought it hard to read -- I can do that though :+1: